### PR TITLE
NAS-132706 / 25.04 / Disallow deletion of the current active user

### PR DESCRIPTION
**Pull request by creatorcary**
`logger.warn` is deprecated, we should replace it.

### CODE REVIEW

**Review by yocalebo — CHANGES_REQUESTED** [review]
We should add a test case

File: src/middlewared/middlewared/plugins/account.py [hunk 1]
```diff
@@ -873,6 +874,9 @@ def do_delete(self, audit_callback, pk, options):
         user = self.middleware.call_sync('user.get_instance', pk)
         audit_callback(user['username'])
 
+        if user['username'] == self.middleware.call_sync('auth.me', app=app)['pw_name']:
```
Comment (anodos325) [inline src/middlewared/middlewared/plugins/account.py:877]: You should be able to extract current user token from `app` without calling auth.me.

File: src/middlewared/middlewared/plugins/account.py [hunk 2]
```diff
@@ -846,7 +846,8 @@ def recreate_homedir_if_not_exists(self, user, group, mode):
                 }).wait_sync(raise_error=True)
 
     @api_method(UserDeleteArgs, UserDeleteResult, audit='Delete user', audit_callback=True)
-    def do_delete(self, audit_callback, pk, options):
+    @pass_app(require=True)
```
Comment (anodos325) [inline src/middlewared/middlewared/plugins/account.py:849]: We probably need to allow REST here. `require` means that we can't have internal callers to `user.delete` which is also probably not intended.

File: src/middlewared/middlewared/plugins/auth.py [hunk 3]
```diff
@@ -931,7 +931,7 @@ async def logout(self, app):
         return True
 
     @api_method(AuthMeArgs, AuthMeResult, authorization_required=False)
-    @pass_app()
+    @pass_app(require=True)
     async def me(self, app):
```
Comment (anodos325) [inline src/middlewared/middlewared/plugins/auth.py:935]: Why are we changing to require=True?

File: src/middlewared/middlewared/plugins/account.py (hunk 2 above)
Comment (creatorcary) [reply src/middlewared/middlewared/plugins/account.py:849]: `require` just means that `app` cannot be `None`, no? We can pass `app` in our internal calls like I do on line 877.

File: src/middlewared/middlewared/plugins/auth.py (hunk 3 above)
Comment (creatorcary) [reply src/middlewared/middlewared/plugins/auth.py:935]: `app` can't be `None`

File: src/middlewared/middlewared/plugins/account.py (hunk 1 above)
Comment (creatorcary) [reply src/middlewared/middlewared/plugins/account.py:877]: Working on a test case now, will get back to this

File: src/middlewared/middlewared/plugins/account.py (hunk 2 above)
Comment (anodos325) [reply src/middlewared/middlewared/plugins/account.py:849]: > require just means that app cannot be None, no?

Right. And there may be valid internal callers for it where we don't want to pass the authenticated credentials.

Comment (anodos325) [reply src/middlewared/middlewared/plugins/account.py:849]: If you don't set `rest=True` then method is unavailable over REST this breaks REST API access to user deletion.

File: src/middlewared/middlewared/plugins/account.py [hunk 4]
```diff
@@ -873,6 +874,9 @@ def do_delete(self, audit_callback, pk, options):
         user = self.middleware.call_sync('user.get_instance', pk)
         audit_callback(user['username'])
 
+        if app and user['username'] == app.authenticated_credentials.user['username']:
```
Comment (anodos325) [inline src/middlewared/middlewared/plugins/account.py:877]: You need to check whether it's a user session as well. See examples in `utils/privilege.py` 
app and app.authenticated_credentials.is_user_session.

**Review by yocalebo — APPROVED** [review]

**Pull request merged**